src/lib.rs

```diff
@@ -0,0 +1,560 @@
+        assert_eq!(tree.nodes[1].height, 0);
+        tree.insert(25);
+        assert_eq!(tree.root, Some(2));
+        assert_eq!(tree.nodes[0].left_child, None);
+        assert_eq!(tree.nodes[0].right_child, None);
+        assert_eq!(tree.nodes[0].parent, Some(2));
+        assert_eq!(tree.nodes[1].left_child, None);
+        assert_eq!(tree.nodes[1].right_child, None);
+        assert_eq!(tree.nodes[1].parent, Some(2));
+        assert_eq!(tree.nodes[2].left_child, Some(1));
+        assert_eq!(tree.nodes[2].right_child, Some(0));
+        assert_eq!(tree.nodes[2].parent, None);
+        assert_eq!(tree.nodes[0].height, 0);
+        assert_eq!(tree.nodes[1].height, 0);
+        assert_eq!(tree.nodes[2].height, 1);
+    }
+
+    #[test]
+    fn right_left_rot() {
+        //   10             15
+        //    \            /  \
+        //    20     =>  10   20
+        //    /
+        //   15
+        let mut tree: SearchTree<u8> = SearchTree::new();
+        tree.insert(10);
+        assert_eq!(tree.root, Some(0));
+        assert_eq!(tree.nodes[0].left_child, None);
+        assert_eq!(tree.nodes[0].right_child, None);
+        assert_eq!(tree.nodes[0].parent, None);
+        tree.insert(20);
+        assert_eq!(tree.root, Some(0));
+        assert_eq!(tree.nodes[0].height, 1);
+        assert_eq!(tree.nodes[0].left_child, None);
+        assert_eq!(tree.nodes[0].right_child, Some(1));
+        assert_eq!(tree.nodes[0].parent, None);
+        assert_eq!(tree.nodes[1].left_child, None);
+        assert_eq!(tree.nodes[1].right_child, None);
+        assert_eq!(tree.nodes[1].parent, Some(0));
+        assert_eq!(tree.nodes[0].height, 1);
+        assert_eq!(tree.nodes[1].height, 0);
+        tree.insert(15);
+        assert_eq!(tree.root, Some(2));
+        assert_eq!(tree.nodes[0].left_child, None);
+        assert_eq!(tree.nodes[0].right_child, None);
+        assert_eq!(tree.nodes[0].parent, Some(2));
+        assert_eq!(tree.nodes[1].left_child, None);
+        assert_eq!(tree.nodes[1].right_child, None);
+        assert_eq!(tree.nodes[1].parent, Some(2));
+        assert_eq!(tree.nodes[2].left_child, Some(0));
+        assert_eq!(tree.nodes[2].right_child, Some(1));
+        assert_eq!(tree.nodes[2].parent, None);
+        assert_eq!(tree.nodes[0].height, 0);
+        assert_eq!(tree.nodes[1].height, 0);
+        assert_eq!(tree.nodes[2].height, 1);
+    }
+
+    #[test]
+    fn random_values() {
+        const LENGTH: usize = 10000;
+        let mut v: Vec<u64> = Vec::with_capacity(LENGTH);
+        for _ in 0..(LENGTH + 1) {
+            v.push(rand::random::<u64>());
+        }
+        let mut tree: SearchTree<u64> = SearchTree::new();
+        for &value in v.iter() {
+            tree.insert(value);
+            // the tree should always be balanced
+            for node_index in 0..tree.nodes.len() {
+                assert!(tree.is_balanced(node_index));
+            }
+        }
+        {
```

Comment:
Is this extra scoping needed to silence the borrow checker?

Reply:
Ah, it used to be needed with my older implementation of the test. It's no longer necessary, I'll remove it